### PR TITLE
Port ContainmentTests

### PR DESCRIPTION
**Pull request by ElizabethOkerio**
### Issues

*This pull request fixes #xxx.*

### Description

This PR ports E2E client tests from depending on WCF services to use Microsoft.AspNetCore.OData. Here is a link to the tests that need porting: https://github.com/OData/odata.net/blob/release-7.x/test/EndToEndTests/Tests/Client/Build.Desktop/ContainmentTest/ContainmentTest.cs

These tests test Containment ODataNet.

### Checklist (Uncheck if it is not completed)

- [ ] *Test cases added*
- [ ] *Build and test with one-click build and test script passed*

### Additional work necessary

*If documentation update is needed, please add "Docs Needed" label to the issue and provide details about the required document change in the issue.*

### CODE REVIEW

**Review by marabooy — APPROVED** [review]

File: test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/Common/Server/Default/DefaultDataSource.cs [hunk 1]
```diff
@@ -756,13 +762,15 @@ public void Initialize()
                         Amount = 1.9,
                         ExperationDate = new DateTimeOffset(new DateTime(2013, 12, 30))
                     },
-                    MyPaymentInstruments =
-                    [
+                    MyPaymentInstruments = new List<PaymentInstrument>()
+                    {
```
Comment (marabooy) [inline test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/Common/Server/Default/DefaultDataSource.cs:766]: question: is there a reason to switch from collection shorthand in tests? non-blocking

File: test/EndToEndTests/Common/Microsoft.OData.Client.E2E.TestCommon/Common/CustomHttpClientRequestMessage.cs [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+// </copyright>
+//------------------------------------------------------------------------------
+
+namespace Microsoft.OData.Client.E2E.TestCommon.Common
+{
+    /// <summary>
+    /// An implementation of IODataRequestMessageAsync that uses an HttpRequestMessage under the covers.
+    /// In ODataLibrary, a message is an abstraction which consists of stream and header interfaces that hides the details of stream-reading/writing.
+    /// </summary>
+    public class CustomHttpClientRequestMessage : IODataRequestMessageAsync, IServiceCollectionProvider
```
Comment (gathogojr) [inline test/EndToEndTests/Common/Microsoft.OData.Client.E2E.TestCommon/Common/CustomHttpClientRequestMessage.cs:14]: Maybe you first merge the other pull request where you made major changes to the classes under _test/EndToEndTests/Common/Microsoft.OData.Client.E2E.TestCommon/Common_

File: test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/Common/Server/Default/DefaultDataModel.cs [hunk 3]
```diff
@@ -202,7 +203,11 @@ public class Company
     public class PublicCompany : Company
     {
         public string? StockExchange { get; set; }
+
+        [Contained]
```
Comment (gathogojr) [inline test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/Common/Server/Default/DefaultDataModel.cs:207]: My question about changing the navigation properties to contained applies. Shouldn't we have a different model for containment tests?

File: test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/Common/Server/Default/DefaultDataSource.cs (hunk 1 above)
Comment (ElizabethOkerio) [reply test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/Common/Server/Default/DefaultDataSource.cs:766]: No reason at all. You think we should use the short-hand?

File: test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/Common/Server/Default/DefaultDataModel.cs (hunk 3 above)
Comment (ElizabethOkerio) [reply test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/Common/Server/Default/DefaultDataModel.cs:207]: This is a common CSDL that is shared by several tests:  https://github.com/OData/odata.net/blob/release-7.x/test/EndToEndTests/Services/TestServices/Service%20References/ODataWCFServiceReference/Reference.cs. I had forgotten to update some of these types as contained or not.. For new tests that test anything that is not in these common csdl, I recommend a new CSDL be created.

File: test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/Common/Server/Default/DefaultDataSource.cs (hunk 1 above)
Comment (ElizabethOkerio) [reply test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/Common/Server/Default/DefaultDataSource.cs]: I've update this to use the shorthand

File: test/EndToEndTests/Common/Microsoft.OData.Client.E2E.TestCommon/Common/CustomHttpClientRequestMessage.cs (hunk 2 above)
Comment (ElizabethOkerio) [reply test/EndToEndTests/Common/Microsoft.OData.Client.E2E.TestCommon/Common/CustomHttpClientRequestMessage.cs]: merged

**Comment by gathogojr** [comment]
nit: If `ContainmentTest` and `ContainmentTestController` files are new files you introduced, perhaps you should align the naming with how you have been doing it in other pull requests, i.e., _ContainmentTest_ => _ContainmentTests_ and _ContainmentTestController_ => _ContainmentTestsController_

File: test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/ContainmentTest/Server/ContainmentTestController.cs [hunk 4]
```diff
@@ -0,0 +1,516 @@
...
+                existingPaymentInstrument.CreatedDate = updatedPaymentInstrument.CreatedDate;
+            }
+
+            var result = account.MyPaymentInstruments.SingleOrDefault(pi => pi.PaymentInstrumentID == paymentInstrumentId);
+
+            return Updated(result);
+        }
+
+
+        [HttpPost("odata/containmenttests/Default.ResetDefaultDataSource")]
```
Comment (gathogojr) [inline test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/ContainmentTest/Server/ContainmentTestController.cs:508]: Is this a non-OData endpoint or how does the `/containmenttests` segment get resolved? Or does it work just okay for a non-bound `ResetDefaultDataSource` action? Can you confirm that it'd work if it the endpoint URI was invoked

Comment (ElizabethOkerio) [reply test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/ContainmentTest/Server/ContainmentTestController.cs]: Yes this works. 

Doing this from the test class works ok: 

```c#

 private void ResetDefaultDataSource()
        {
            var actionUri = new Uri(_baseUri + "containmenttests/Default.ResetDefaultDataSource", UriKind.Absolute);
            _context.Execute(actionUri, "POST");
        }

```

**Review by gathogojr — APPROVED** [review]

**Pull request merged**